src/Common/MemoryBufferWriter.cs

```diff
@@ -146,43 +146,49 @@ public override Task CopyToAsync(Stream destination, int bufferSize, Cancellatio
 
         private void EnsureCapacity(int sizeHint)
         {
-            // TODO: Use sizeHint
-            if (_currentSegment != null && _position < _currentSegment.Length)
+            // This does the Right Thing. It only subtracts _position from the current segment length if it's non-null.
+            // If _currentSegment is null, it returns 0.
+            var remainingSize = _currentSegment?.Length - _position ?? 0;
+
+            // If the sizeHint is 0, any capacity will do
+            // Otherwise, the buffer must have enough space for the entire size hint, or we need to add a segment.
+            if ((sizeHint == 0 && remainingSize > 0) || (sizeHint > 0 && remainingSize >= sizeHint))
             {
                 // We have capacity in the current segment
                 return;
             }
 
-            AddSegment();
+            AddSegment(sizeHint);
         }
 
-        private void AddSegment()
+        private void AddSegment(int sizeHint = 0)
         {
             if (_currentSegment != null)
             {
                 // We're adding a segment to the list
-                if (_fullSegments == null)
+                if (_completedSegments == null)
                 {
-                    _fullSegments = new List<byte[]>();
+                    _completedSegments = new List<BufferSegment>();
                 }
 
-                _fullSegments.Add(_currentSegment);
+                _completedSegments.Add(new BufferSegment(_currentSegment, _position));
```

Comment:
Add a comment explaining why/when the position might be less than the complete segment length